### ISSUE TITLE
On Linux aarch64, java command fails with -Xgc:enableArrayletDoubleMapping option

### ISSUE THREAD

**Issue by avermeer** [eclipse-openj9/openj9#18104]
Java -version output
--------------------
`openjdk version "17.0.8.1" 2023-08-24 
IBM Semeru Runtime Open Edition 17.0.8.1 (build 17.0.8.1+1) 
Eclipse OpenJ9 VM 17.0.8.1 (build openj9-0.40.0, JRE 17 Linux aarch64-64-Bit Compressed References 20230824_507 (JIT enabled, AOT enabled) 
OpenJ9 - d12d10c9e 
OMR - e80bff83b 
JCL - 8ecf238a124 based on jdk-17.0.8.1+1)`

Summary of problem
------------------

The 'java' command fails to execute `-Xgc:enableArrayletDoubleMapping` option specifically on Linux aarch64, while the same option with exact same OpenJ9 (or IBM Semeru) version works fine on Linux x64.

Looking at https://github.com/eclipse-openj9/openj9/issues/8321 makes me thing that -Xgc:enableArrayletDoubleMapping option is supposed to work with OpenJ9, no matter x64 or aarch64.

Please note that I have tried adding -XX:+IgnoreUnrecognizedXXColonOptions option, but that does not solve this issue. 

An option to let OpenJ9 aarch64 ignore this option if it's not suitable for this processor architecture would be appreciated, or a patch otherwise...

Diagnostic files
----------------

On a Linux aarch64 machine:

`[root@i-6c8e4bf6 bin]# ./java -version 
openjdk version "17.0.8.1" 2023-08-24 
IBM Semeru Runtime Open Edition 17.0.8.1 (build 17.0.8.1+1) 
Eclipse OpenJ9 VM 17.0.8.1 (build openj9-0.40.0, JRE 17 Linux aarch64-64-Bit Compressed References 20230824_507 (JIT enabled, AOT enabled) 
OpenJ9 - d12d10c9e 
OMR - e80bff83b 
JCL - 8ecf238a124 based on jdk-17.0.8.1+1) 
[root@i-6c8e4bf6 bin]# ./java -Xgc:enableArrayletDoubleMapping -version 
JVMJ9GC040E Unknown option: enableArrayletDoubleMapping 
JVMJ9VM015W Initialization error for library j9gc29(2): Failed to initialize; unable to parse command line 
Error: Could not create the Java Virtual Machine. 
Error: A fatal exception has occurred. Program will exit.`

With exact same Java version on Linux x64, it works:

`[root@ip-172-31-35-204 bin]# ./java -version 
openjdk version "17.0.8.1" 2023-08-24 
IBM Semeru Runtime Open Edition 17.0.8.1 (build 17.0.8.1+1) 
Eclipse OpenJ9 VM 17.0.8.1 (build openj9-0.40.0, JRE 17 Linux amd64-64-Bit Compressed References 20230824_549 (JIT enabled, AOT enabled) 
OpenJ9 - d12d10c9e 
OMR - e80bff83b 
JCL - 8ecf238a124 based on jdk-17.0.8.1+1) 
[root@ip-172-31-35-204 bin]# ./java -Xgc:enableArrayletDoubleMapping -version 
openjdk version "17.0.8.1" 2023-08-24 
IBM Semeru Runtime Open Edition 17.0.8.1 (build 17.0.8.1+1) 
Eclipse OpenJ9 VM 17.0.8.1 (build openj9-0.40.0, JRE 17 Linux amd64-64-Bit Compressed References 20230824_549 (JIT enabled, AOT enabled) 
OpenJ9 - d12d10c9e 
OMR - e80bff83b 
JCL - 8ecf238a124 based on jdk-17.0.8.1+1)`

**Comment by avermeer** [comment]
Please note that the same issue lies with IBM Semeru 17.0.7 too, so it's not a regression in 17.0.8.1...

**Comment by pshipton** [comment]
@dmitripivkine fyi

**Comment by dmitripivkine** [comment]
@avermeer Would you please provide Linux kernel versions?

**Comment by avermeer** [comment]
Here it is:

`root@i-6c8e4bf6 ~]# uname -r 
4.18.0-348.20.1.el7.aarch64`

**Comment by dmitripivkine** [comment]
`J9VM_GC_ENABLE_DOUBLE_MAP` is not set for Linux aarch64, so feature code isn't included to the build. I don't know is it just missed or there is some problem with it. I guess this feature has not been even tested on aLinux.
Please note Arraylet Double Mapping feature is applicable for `-Xgcpolicy:Balanced` only. Any other GC policy would not be beneficial from its enabling.

**Comment by avermeer** [comment]
Understand that this only applies to `-Xgcpolicy:Balanced`.

Is there any way to by-pass the command-line parsing error (like ignore unknown gc options) ?

**Comment by dmitripivkine** [comment]
@avermeer Would you please let me know if you are planning to run JVM with GC policy Balanced on aLinux soon? We are short in hands but if you need fix for this earlier I can try to prioritize this item.
Also please note that an alternative solution coming https://github.com/eclipse-openj9/openj9/issues/14005
It is on late stages of testing and is planning to be released later in 2023 or early in 2024. This feature is going to make Arraylet Double Mapping obsolete.

**Comment by avermeer** [comment]
We are already using `-Xgcpolicy:balanced` with our app on Linux x64, and we were expecting that with same options, our app would start on Linux aarch64.

At the moment, and even if that sounds a low requirement, I'm expecting a way to by-pass the fact that JVM exits when `-Xgc:enableArrayletDoubleMapping` option is there. 

For the longer term, yes a proper way to support balanced policy or equivalent gc policy on aarch64 would help.

**Comment by dmitripivkine** [comment]
ok, I understand. Please let me see what I can do to ignore option silently.

**Comment by dmitripivkine** [comment]
improved:
Head: https://github.com/eclipse-openj9/openj9/pull/18109
0.41.0 release: https://github.com/eclipse-openj9/openj9/pull/18114

**Comment by avermeer** [comment]
@dmitripivkine thanks for the pull request, any Linux aarch64 build which I could use to test your fix?

**Comment by pshipton** [comment]
All we have atm is the latest head stream build.
https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Build_JDK17_aarch64_linux_Nightly/546/OpenJ9-JDK17-aarch64_linux-20230911-230753.tar.gz

**Comment by avermeer** [comment]
Hello,

I just tried with Milestone 1 builds for the Oct release, and I am afraid that the fix is not in this milestone release on Linux ARM (using RedHat 8.7):

```
root@i-adde4c6a bin]# ./java -version 
openjdk version "17.0.9" 2023-10-17 
IBM Semeru Runtime Open Edition 17.0.9.0-m1 (build 17.0.9+5) 
Eclipse OpenJ9 VM 17.0.9.0-m1 (build v0.41.0-release-87d042a68, JRE 17 Linux aarch64-64-Bit Compressed References 20231017_515 (JIT enabled, AOT enabled) 
OpenJ9 - 87d042a68 
OMR - fa7b6ddc7 
JCL - 7e7cdc3f699 based on jdk-17.0.9+5) 
```
and with the `-Xgc:enableArrayletDoubleMapping` option we still get the error:

```
[root@i-adde4c6a  bin]# ./java -Xgc:enableArrayletDoubleMapping -version 
JVMJ9GC040E Unknown option: enableArrayletDoubleMapping 
JVMJ9VM015W Initialization error for library j9gc29(2): Failed to initialize; unable to parse command line 
Error: Could not create the Java Virtual Machine. 
Error: A fatal exception has occurred. Program will exit. 
```

Any idea why @dmitripivkine 's fix is not in this Milestone 1 release of OpenJ9 0.41.0 ?

What's the blocker to have it in OpenJ9 0.41.0 ?

Thanks,
Alexandre

**Comment by avermeer** [comment]
Also, I made a test with https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Build_JDK17_aarch64_linux_Nightly/554/OpenJ9-JDK17-aarch64_linux-20230921-230833.tar.gz, and this one has the expected fix, as this test shows:

```
[root@i-adde4c6a  bin]# ./java -Xgc:enableArrayletDoubleMapping -version 
openjdk version "17.0.9-internal" 2023-10-17 
OpenJDK Runtime Environment (build 17.0.9-internal+0-adhoc..BuildJDK17aarch64linuxNightly) 
Eclipse OpenJ9 VM (build master-1d693b194ba, JRE 17 Linux aarch64-64-Bit Compressed References 20230921_554 (JIT enabled, AOT enabled) 
OpenJ9 - 1d693b194ba 
OMR - 1b52192f57f 
JCL - eda5e2c9b99 based on jdk-17.0.9+7) 
```

So it looks like the fix was not ported in M1 release of OpenJ9 0.41.0, can it be ported to M2 please?

Thanks,
Alex

**Comment by pshipton** [comment]
Yes, it will be in M2 as per https://github.com/eclipse-openj9/openj9/pull/18114, the change was merged after the M1 builds occurred.

**Comment by avermeer** [comment]
@pshipton thank you very much: it is a relief to know that this will eventually be part of 17.0.9 final !

**Comment by avermeer** [comment]
Hello,

I tested the fix with Milestone 2 build for the Oct release (to be more precise, using this binary archive: https://github.com/ibmruntimes/semeru17-binaries/releases/download/jdk-17.0.9%2B7_openj9-0.41.0-m2/ibm-semeru-open-jre_aarch64_linux_17.0.9_7_openj9-0.41.0-m2.tar.gz).

Good news is is that now it works as expected, i.e., this command-line no longer fails:

```
./java -Xgc:enableArrayletDoubleMapping -version 
openjdk version "17.0.9" 2023-10-17 
IBM Semeru Runtime Open Edition 17.0.9.0-m2 (build 17.0.9+7) 
Eclipse OpenJ9 VM 17.0.9.0-m2 (build v0.41.0-release-3ede3b5a4, JRE 17 Linux aarch64-64-Bit Compressed References 20231017_534 (JIT enabled, AOT enabled) 
OpenJ9 - 3ede3b5a4 
OMR - fa7b6ddc7 
JCL - 4024096e8d2 based on jdk-17.0.9+7) 
```

Thank you very much to all who contributed to this fix to make its path into OpenJ9 0.41.0 !

Kind regards,
Alexandre
